core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -606,56 +620,73 @@ public SourceDistributionTaskSource(
                 }
             }
 
-            Optional<List<Object>> tableExecuteSplitsInfo = splitSource.getTableExecuteSplitsInfo();
+            verify(splitSource.isFinished() || !result.isEmpty());
```

Comment:
This doesn't feel necessary as it's the only condition that can exit the loop?

Reply:
I agree. This is kida obvious. Droping.